Hw02 express

vladdubrovskyi opened the pull request:
Добрий вечір. Прохання перевірити. Дякую!

vladdubrovskyi committed:
set up routes and functions to work with db

vladdubrovskyi committed:
adds validation

vladdubrovskyi committed:
fix bugs

KhYulian commented on controllers/controllers.js:
```diff
@@ -0,0 +1,54 @@
+const contactsDataBase = require("../models/contacts");
```
зробіть деструктуризацію

vladdubrovskyi replied:
Готово

KhYulian commented on controllers/controllers.js:
```diff
@@ -0,0 +1,54 @@
+  const body = req.body;
+  if (!body) {
+    return next(HttpError(400, "missing fields"));
+  }
```
для цього використовуйте Joi

vladdubrovskyi replied:
Готово

KhYulian commented on helpers/index.js:
```diff
@@ -0,0 +1,20 @@
+  };
+}
+
+function HttpError(status, message) {
+  const error = new Error(message);
+  error.status = status;
+  return error;
+}
+
```
використовуйте класи

vladdubrovskyi replied:
готово

KhYulian commented on routes/api/contacts.js:
```diff
@@ -1,25 +1,39 @@
+    "any.required": "you should provide email!!",
+  }),
+  phone: Joi.string().required().min(6),
+});
```
Виенсіть const  ContactSchema в окремий файл

vladdubrovskyi replied:
Готово

KhYulian commented on routes/api/contacts.js:
```diff
@@ -1,25 +1,39 @@
-module.exports = router
+router.put(
+  "/:contactId",
+  validateBody(ContactSchema),
```
створіть окрему Joi схему для оновлення контакту

vladdubrovskyi replied:
Готово

KhYulian reviewed (changes requested):
Потрібне доопрацювання

KhYulian commented:
Відповідайте на коменти по мірі виправлення :)

vladdubrovskyi committed:
fix 2 bugs

Зробив деструктуризацію contactsDataBase  та виніс Schema в окремий файл

vladdubrovskyi committed:
fix another bugs

Додає окрему схему валідації для оновлення контакту.
Перевірка наявності хоча б одного поля req,body при put-запиті здійснюється через Joi

vladdubrovskyi committed:
adds class for errors

KhYulian commented on helpers/index.js:
```diff
@@ -0,0 +1,32 @@
+
+function hanldeError(message, status) {
+  throw new MyError(message, status);
+}
```
дуже сумнівна практика. Що заважає просто написати  throw new MyError(message, status)  де потрібно

vladdubrovskyi replied:
Виправив

KhYulian commented on controllers/controllers.js:
```diff
@@ -0,0 +1,56 @@
+  const { contactId } = req.params;
+  const contact = await updateContact(contactId, req.body);
+  if (!contact) {
+    return next(hanldeError("not found", 404));
```
hanldeError - не використовуйте

vladdubrovskyi replied:
Готово

KhYulian commented on schema/schema.js:
```diff
@@ -0,0 +1,24 @@
+  email: Joi.string().email(),
+  phone: Joi.string().min(6),
+})
+  .required()
```
.required() для оновлення не потрібен. Достатньо min(1)

vladdubrovskyi replied:
Виправив

KhYulian commented on helpers/index.js:
```diff
@@ -0,0 +1,32 @@
+  const error = new Error(message);
+  error.status = status;
+  return error;
+}
```
якщо використовуте MyError, то ця функ. вам вже не потрібна

vladdubrovskyi replied:
Видалив

vladdubrovskyi committed:
updates class for errors and fixes schema

KhYulian reviewed (approved):
Чудова робота :)
